fix(ios): use native NSTimer to avoid thread deadlocks

janvennemann opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-26391

**Optional Description:**
Our KrollTimer implementation is using his own thread to wait until the timer is fired. This causes thread deadlocks under iOS 12. Using iOS native timer to operate directly on the run loop to try mitigate this problem.

The implementation is somewhat naive and may require some adjustments for proper memory management of the timer callbacks.

janvennemann committed:
use native timers

janvennemann committed:
only test with JSC and run on mainthread

janvennemann committed:
add forward declaration

janvennemann committed:
release objects

janvennemann commented on iphone/Classes/KrollContext.m:
```diff
@@ -1529,3 +1536,59 @@ - (void)dealloc
+  NSNumber *timerIdentifier = [NSNumber numberWithUnsignedInteger:[[[[NSUUID alloc] init] autorelease] hash]];
+  NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:interval repeats:repeats block:^(NSTimer * _Nonnull timer) {
+    [callbackFunction callWithArguments:nil];
+    [callbackFunction release];
```
@vijaysingh-axway not sure if it's a good idea to directly release the callback again here. Any thoughts?

vijaysingh-axway replied:
@janvennemann 
1. In case of setTimeout we should remove timer from self.timers array.
2. Is there any reason you have retained in callbackFunction. There is neither need of retain in callbackFunction nor release. Above code is crashing for me.
Test case -
`setTimeout(function(e){
      Ti.API.info('timeout called');
}, 2000);

setTimeout(function(e){
      Ti.API.info('timeout called second time');
}, 4000);

setInterval(function(e){
      Ti.API.info('time interval 1 called');
}, 3000);

setInterval(function(e){
      Ti.API.info('time interval 2 called');
}, 4000);`

build commented:
<!--
  2 failure:  :disappointed_rel..., :microscope: Ther...
  0 warning: 
  1 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/janvennemann/titanium_mobile/blob/timer-refactor/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.5.0 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10401-UEH4EG3QUGV3PUF2GRLI2T256KXVLLDKVKZ6EY26WECF5NYXLBQQ
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10401-UEH4EG3QUGV3PUF2GRLI2T256KXVLLDKVKZ6EY26WECF5NYXLBQQ/tests/Resources/ti.network.httpclient.addontest.js
  11:5  warning  'should' is assigned a value but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)


Running "checkFormat:ios" (checkFormat) task
Fatal error: Error: Formatting incorrect on "iphone/Classes/KrollContext.m", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='37058' length='1'>&#10;            </replacement>
<replacement offset='37333' length='1'>&#10;            </replacement>
<replacement offset='49356' length='1'>&#10;</replacement>
<replacement offset='49545' length='6'>&#10;&#10;  </replacement>
<replacement offset='49731' length='6'>&#10;&#10;  </replacement>
<replacement offset='50039' length='1'>&#10;</replacement>
<replacement offset='50294' length='1'>&#10;</replacement>
<replacement offset='50610' length='1'>&#10;                                                   </replacement>
<replacement offset='50626' length='1'>&#10;                                                     </replacement>
<replacement offset='50644' length='1'></replacement>
<replacement offset='50662' length='5'>&#10;                                                       </replacement>
<replacement offset='50708' length='5'>&#10;                                                       </replacement>
<replacement offset='50740' length='3'>&#10;                                                     </replacement>
<replacement offset='50909' length='1'>&#10;</replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

joshualambert commented:
I've been testing with this code change on my app. Super stable and fast now on iOS12. Fantastic work @janvennemann!

vijaysingh-axway commented on iphone/Classes/KrollContext.m:
```diff
@@ -1529,3 +1536,59 @@ - (void)dealloc
 }
 
 @end
+
```
Please wrap whole class implementation  inside macro
#if defined(USE_JSCORE_FRAMEWORK) && !defined(TI_USE_KROLL_THREAD)

vijaysingh-axway commented on iphone/Classes/KrollContext.m:
```diff
@@ -1177,10 +1179,15 @@ - (void)main
   [kroll release];
   TiStringRelease(prop);
 
+#if defined(USE_JSCORE_FRAMEWORK) && !defined(TI_USE_KROLL_THREAD)
```
I think the problem is with JSCore only irrespective of main thread or kroll thread. If so we should update the # if defined statement to - 
#if defined(USE_JSCORE_FRAMEWORK)

janvennemann committed:
remove unnecessary retain/release

remove unused JSContext parameter
format code

janvennemann committed:
fix formatting

sgtcoolguy committed:
Merge branch 'master' into timer-refactor

sgtcoolguy committed:
Fix formatting

janvennemann commented on iphone/Classes/KrollContext.m:
```diff
@@ -1177,11 +1179,18 @@ - (void)main
+  [JSTimerManager initializeInContext:jsContext];
+#else
+  [self bindCallback:@"setTimeout"
+            callback:&SetTimeoutCallback];
```
No idea why it suddenly wants to format it like this. Total non-sense if you ask me. If you remove the `#ifdef` the break is not required anymore.

janvennemann committed:
pass remaining args to callback

janvennemann committed:
exclude legacy timers in jsc

janvennemann committed:
move interface declaration to header and add docs

janvennemann committed:
exclude unused methods when using jsc

janvennemann committed:
fix leak and array mutations

janvennemann committed:
invalidate timers

janvennemann committed:
create a timer manager per KrollContext

janvennemann committed:
fix formatting and ad missing endif

janvennemann commented:
@vijaysingh-axway i addressed your suggested changes and also reworked a few bits of this. Every `KrollContext` now has it's own timer manager instance instead of working with static class methods. I also added some clean up code and excluded some of the legacy timer code with macros if using jscore.

vijaysingh-axway committed:
Merge branch 'master' into timer-refactor

vijaysingh-axway commented on iphone/Classes/KrollContext.m:
```diff
@@ -1407,6 +1422,9 @@ - (void)main
   }
 
+#ifdef USE_JSCORE_FRAMEWORK
+  [JSTimerManager invalidateAllTimers];
```
It should be -
[timerManager invalidateAllTimers];

sgtcoolguy committed:
Merge branch '7_5_X' into timer-refactor

janvennemann committed:
do not use new timers on kroll thread

If running on kroll thread native timers never fire because kroll thread main method blocks the run loop with an infinite loop

janvennemann committed:
address review comments

janvennemann commented on iphone/Classes/KrollContext.m:
```diff
@@ -1529,3 +1548,80 @@ - (void)dealloc
+  NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:interval
+                                                   repeats:shouldRepeat
+                                                     block:^(NSTimer *_Nonnull timer) {
+                                                       [callbackFunction callWithArguments:callbackArgs];
```
I tested with Kroll thread and timers are not properly firing here. I think it's because we are permanently blocking the run loop with our infinite while loop. For now i only made the new timer management available when using jscore *and* running on main thread.

janvennemann committed:
wait for timer callback to finish on kroll thread

build commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-default;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10401/18/artifact/dist/mobilesdk-7.5.0.v20181030124741-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

vijaysingh-axway commented on package-lock.json:
```diff
@@ -477,7 +477,7 @@
     "babel-helper-is-nodes-equiv": {
       "version": "0.0.1",
-      "resolved": "http://registry.npmjs.org/babel-helper-is-nodes-equiv/-/babel-helper-is-nodes-equiv-0.0.1.tgz",
+      "resolved": "https://registry.npmjs.org/babel-helper-is-nodes-equiv/-/babel-helper-is-nodes-equiv-0.0.1.tgz",
```
I think this file should not be commit.

vijaysingh-axway commented on iphone/Classes/KrollTimer.m:
```diff
@@ -108,7 +108,7 @@ - (void)main
       TiThreadPerformOnMainThread(^{
         [kroll invokeOnThread:self method:@selector(invokeWithCondition:) withObject:invokeCond condition:nil];
       },
-          NO);
```
@janvennemann Please add KrollTimer.m and KrollTimer.h file's under macro.

sgtcoolguy committed:
Merge branch '7_5_X' into timer-refactor

hansemannn commented on iphone/Classes/KrollContext.h:
```diff
@@ -175,3 +183,29 @@ KrollContext *GetKrollContext(TiContextRef context);
+/**
+ * Handles creating and clearing timers when running with JavaScriptCore
+ */
+@interface JSTimerManager : NSObject
```
Small peak: This may clash with internal API's some day, since it uses the JavaScriptCore framework prefix. To be save, naming it like `TiTimerManager` or `TiJSTimerManager` may help.

hansemannn commented on iphone/Classes/KrollContext.h:
```diff
@@ -175,3 +183,29 @@ KrollContext *GetKrollContext(TiContextRef context);
+/**
+ * Map of timer identifiers and the underlying native NSTimer.
+ */
+@property (nonatomic, strong) NSMutableDictionary<NSNumber *, NSTimer *> *timers;
```
If timers are supposed to be unique identified, an `NSMutableSet` may be more appropriate.

janvennemann replied:
Since i need to access timers by their id in various places a set is somewhat cumbersome to work with in that use case.

hansemannn commented on iphone/Classes/KrollContext.m:
```diff
@@ -1529,3 +1544,84 @@ - (void)dealloc
+
+  self.timers = [NSMutableDictionary new];
+
+  int (^setInterval)(void) = ^() {
```
Any reason to prefer the C `int` type vs. ObjC `NSInteger`?

hansemannn commented on iphone/Classes/KrollContext.m:
```diff
@@ -1529,3 +1544,84 @@ - (void)dealloc
+
+- (void)clearIntervalWithIdentifier:(int)identifier
+{
+  NSNumber *timerIdentifier = [NSNumber numberWithInteger:identifier];
```
In general: All `[NSNumber numberWith****]` convenience constructors should use the boxed version of it (`@(identifier)`) since Objective C 2.0.

janvennemann committed:
address various review comments

simplify timer identifier generation
use NSNumber literals
exclude KrollTimer if not used
avoid possible naming conflict

sgtcoolguy committed:
Merge branch '7_5_X' into timer-refactor
